### DISCUSSION TITLE
Use of jax.lax.stop_gradient in nn.softmax

### DISCUSSION THREAD

**Discussion by amishra791** [jax-ml/jax#24935] [Q&A] [answered]
Hi, I was taking a look at the softmax [function](https://github.com/jax-ml/jax/blob/main/jax/_src/nn/functions.py#L592) and it's deprecated [version](https://github.com/jax-ml/jax/blob/main/jax/_src/nn/functions.py#L611). I noticed that in the deprecated version has `jax.lax.stop_gradient`. What is the use of this function here, and why is it not used in the most recent implementation of softmax?

**Comment by jakevdp** [comment, marked as answer]
The reason for this is that the non-deprecated softmax has a custom JVP, and so we don't differentiate through the implementation, which makes the `stop_gradient` call unnecessary.

**Reply by amishra791** [reply]
Thanks! Could you explain more as to why we needed the `stop_gradient` in the deprecated version?

**Reply by jakevdp** [reply]
See https://stackoverflow.com/questions/74822097/purpose-of-stop-gradient-in-jax-nn-softmax
